Sync payload

bausmeier opened the pull request.

bausmeier committed:
Create a p2p module

To implement the interfaces and classes which provide the core
functionality for syncing resources from peer to peer.

bausmeier committed:
Add SyncPayload class

bausmeier commented:
@trevorgowing This is what I had in progress so far. Please take a look and let me know what you think. I still just want to clean things up and write some more tests.

bausmeier committed:
Apply spotless formatting

trevorgowing committed:
Add dependency on kotlin coroutines core to p2p

trevorgowing commented:
@bausmeier In your last comment you mentioned that you wanted to clean things up and add some tests. Are we still waiting for that or can I merge this PR?

bausmeier commented:
I think it's good enough to merge as-is, unless you have any feedback on the code or the test?

trevorgowing commented on p2p/src/main/java/com/google/android/fhir/p2p/SyncPayload.kt:
```diff
@@ -0,0 +1,80 @@
+@Serializable
+sealed class RemoteChange {
+  /** The timestamp at which the change took place. */
+  abstract val timestamp: String
```
What was the thinking with making the timestamp a string? It would be serialized to a string but do we not want to type it as a `DateTime`?

bausmeier replied:
I stuck with the same type as what's used by the `LocalChangeEntity` so that it's consistent between the two. I think it's safest not to enforce a particular format here if it's not also used there.

trevorgowing commented on p2p/src/test/java/com/google/android/fhir/p2p/SyncPayloadTest.kt:
```diff
@@ -0,0 +1,58 @@
+@RunWith(BlockJUnit4ClassRunner::class)
+class SyncPayloadTest {
+  @Test
+  fun encodeToStringEncodesThePayloadToAJsonString() {
```
Do you not think that it is worth adding a decode test too?

bausmeier replied:
Ah yes, that's the other test I wanted to add. Will do it now quickly.

bausmeier replied:
I've added a test for decoding the payloads in https://github.com/decode-development/android-fhir/pull/1/commits/3e4804e45deaca20ef623525e3de2fcd16fbc27d.

trevorgowing commented on p2p/src/main/java/com/google/android/fhir/p2p/SyncPayload.kt:
```diff
@@ -0,0 +1,80 @@
+@Serializable
+data class SyncPayload(
+  /** The current version of the resource after all of the changes have been applied. */
+  @Serializable(with = FhirResourceSerializer::class) val resource: IBaseResource,
```
Is `IBaseResource` the level that we want to hook into the hapi model. `engine` uses [Resource](https://github.com/hapifhir/org.hl7.fhir.core/blob/master/org.hl7.fhir.r4/src/main/java/org/hl7/fhir/r4/model/Resource.java) throughout?

bausmeier replied:
I don't think that we can go any higher because `IParser.parseResource` returns `IBaseResource` unless we specify the resource type which we don't know ahead of time.

bausmeier committed:
Encode all expressions as JSON

To make sure that special characters are dealt with consistently.

trevorgowing commented:
Okay. Let's see how it goes. I think we may run into issues here though.

On Wed, 10 Nov 2021, 11:55 Brett Ausmeier, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In p2p/src/main/java/com/google/android/fhir/p2p/SyncPayload.kt
> <https://github.com/decode-development/android-fhir/pull/1#discussion_r746426003>
> :
>
> > +import ca.uhn.fhir.context.FhirContext
> +import kotlinx.serialization.KSerializer
> +import kotlinx.serialization.SerialName
> +import kotlinx.serialization.Serializable
> +import kotlinx.serialization.descriptors.PrimitiveKind
> +import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
> +import kotlinx.serialization.descriptors.SerialDescriptor
> +import kotlinx.serialization.encoding.Decoder
> +import kotlinx.serialization.encoding.Encoder
> +import org.hl7.fhir.instance.model.api.IBaseResource
> +
> +/** The payload for syncing changes between peers. */
> ***@***.***
> +data class SyncPayload(
> +  /** The current version of the resource after all of the changes have been applied. */
> +  @Serializable(with = FhirResourceSerializer::class) val resource: IBaseResource,
>
> I don't think that we can go any higher because IParser.parseResource
> returns IBaseResource unless we specify the resource type which we don't
> know ahead of time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/decode-development/android-fhir/pull/1#discussion_r746426003>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGS5YU4YP22QDC3JQDP5VDULI6QDANCNFSM5HSZWCAQ>
> .
>

bausmeier committed:
Test decoding of sync payloads

Make sure that the resources are correctly deserialized.

bausmeier committed:
Ensure resource is deserialized to the right type

The engine module deals with `Resource`s rather than `IBaseResource`s,
so checking the instance type needs to be possible.
